p4-16/spec/P4-16-spec.mdk

```diff
@@ -6007,6 +6007,19 @@ Figure [#fig-subparser] shows a diagram of this process.
 Note that since P4 requires declarations to precede uses, it is impossible to
 create recursive (or mutually recursive) parsers.
 
+When a parser is instantiated, all its local declarations of stateful instantiations are evaluated recursively.
+Each instantiation of a parser will have a unique set of value sets, extern objects, and inner parser instances.
```

Comment:
I will leave it to other LDWG members to decide whether the meaning is clear to them or not, but I wanted to point out that you can pass as compile-time parameters instances of objects like value sets, extern objects, and/or parser instances, and the instantiated parser can then use those pre-existing instances, as well as creating new instances of those things during the parser's instantiation.

Thus the statement "Each instantiation of a parser will have a unique set of value sets, extern objects, and inner parser instances." is true if all of those things are instantiated within the parser's declaration, but false if you pass pre-existing instances to the parser's instantiation.

The same applies to controls, too.

Similarly the statement below "invoking a parser instance twice is different from invoking two parser instances each one" can be true or false depending upon how the two parser instances are declared and instantiated.  For example, if there are NO local instantiations inside the parser declaration, and it ONLY invokes value sets, extern objects, and inner parser instances that were passed to it at instantiation time, you can have two different parser instances that DO update the same stateful objects, regardless of which parser you invoke at run time.

I think the "heart" of the motivation for these changes is that local variables behave differently than local instantiations of other things, and I'm perfectly fine with clarifying that behavior if it is unclear.  I am simply hoping we don't add other statements that might be confusing or subtly wrong in the process.

Reply:
Hi Andy! Thanks for the review. I totally agree with your concern that no more confusion should be created during the process of solving one. Just my opinion on this specific issue: I think it should be clear by saying it is the "local declarations of stateful instantiations" that are evaluated recursively. The order of evaluation is explained in section [17.2. Compile-time Evaluation](https://p4.org/p4-spec/docs/P4-16-v1.2.2.html#sec-ct-eval), which separates  constructor arguments from local instantiation. But that is just my opinion, and we can tweak the wording here to make things more clear.